### ISSUE TITLE
NEED TESTERS: fallback to interactive root picker if no root directory detected

### ISSUE THREAD

**Issue by mjlbach** [neovim/nvim-lspconfig#1028]
There are basically three usability complaints with lspconfig

1. we don't automatically provide language servers (solved by lspinstall)
2. we don't set up autocompletion + snippet for you (solved by nvim-compe + luasnip/vnsip)
3. Occasionally servers fail to start

3 is because lspconfig tries to find the correct root directory for the project, many language servers are picky about this, but some will still provide some utility without the correct directory.

I'm going to add a fallback hook that can be customized to either:
1. Prompts you interactively to choose the root directory or
2. Prints a warning that no non-parent root pattern has been detected, and tries to start the server in the parent directory

WIP:
![image](https://user-images.githubusercontent.com/13316262/125319558-bc59e080-e2ef-11eb-9092-8636d7a9781a.png)

**Comment by mizlan** [comment]
I think an lsp-mode style prompt is better for usability and troubleshooting. Granted, users will have to (be patient enough to) read through the different options selected in the prompt.

**Comment by mjlbach** [comment]
We got a lot of complaints last time I put an interactive prompt, so mixed.

**Comment by mjlbach** [comment]
Please test #1066 

I had to make sweeping changes to root directory detection (mainly disabling working directory and parent directory fallback) to support this change.

You will be able to override the fallback behavior, but I think this is generally a better default for everyone.

**Comment by dundargoc** [comment]
I can't get the interactive root directory picker to trigger/start, but the root directory isn't being detected either. At least I don't think so, maybe I missed something. It only works when I create a softlink build/.....command.json (something like that) to root manually.

**Comment by mjlbach** [comment]
That's strange, if you `mkdir test` && `cd test` && `touch main.c` && `nvim main.c`, it immediately pops up for me. If you run `:LspInfo` do you see a root? The server can still detect a root and fail to (appropriately) start because compile_command.json is missing. In the event *any* root is detected, lspconfig will never show you the popup.

**Comment by mizlan** [comment]
You should make it more obvious what this is for, i.e., the popup should mention LSP somewhere. This is important because "Ignore this file" has no context otherwise (you may also want to reword it to "Don't enable LSP on this file" or something of the sort).

**Comment by mizlan** [comment]
Also, would there be a way to use ShaDa or some other persistent method to remember a user's decision? I am repeatedly asked for root dir on opening Neovim when I have already chosen it previously. If you were to do that, you would additionally need a way to manually configure, for example if the user changes their mind.

**Comment by mjlbach** [comment]
> Also, would there be a way to use ShaDa or some other persistent method to remember a user's decision? I am repeatedly asked for root dir on opening Neovim when I have already chosen it previously. If you were to do that, you would additionally need a way to manually configure, for example if the user changes their mind.

Yes, but I'm not going to implement this yet until we have workspace support, at which point it will be written to a project local json file.

**Comment by WhyNotHugo** [comment]
I don't think that dropping existing fallbacks in favour of the interactive picker should be the out-of-the-box experience.

The existing fallbacks work pretty solidly. Anyone _wanting_ a picker at runtime canconfigure that manually (and disable the current mechanism), not have that as the default, that would be incredibly annoying.

I'm not a fan of dropping something that works automatically in favour of something that needs to be done manually, at very inconvenient times, over and over by the user.

**Comment by mjlbach** [comment]
You will almost never see the interactive picker if you've configured your project correctly and if the settings have same root directory detection.

**Comment by lithammer** [comment]
Gotta say I'm with @WhyNotHugo on this. Not all files are part of projects. And I definitely don't want a popup every time I edit those.

**Comment by mjlbach** [comment]
In that case it's extremely easy to override the default behavior with a fallback to util.path.dirname which I will document.

As it stands, people are putting util.path.dirname as a fallback for everything, which is not in compliance with the specification and will result in broken behavior (servers will fail to index the whole project) and getting confused when language servers do not work correctly. The root of the project is not/should not be the parent directory of the current file. We can special case a few (EFM, jsonls, for example).

**Comment by mizlan** [comment]
What about:

If a project root is not detected, fail with a message in the echo area telling the user to start it manually (e.g., `LSP failed to autostart, pick a root directory with :LspStart`), and have `:LspStart` display the interactive menu as a fallback?

**Comment by mjlbach** [comment]
> What about:
> 
> If a project root is not detected, fail with a message in the echo area telling the user to start it manually (e.g., `LSP failed to autostart, pick a root directory with :LspStart`), and have `:LspStart` display the interactive menu as a fallback?

That would add an additional step over the behavior of the PR, making hugo's conern worse.

**Comment by WhyNotHugo** [comment]
> Op 2021-07-21 om 19:28 heeft Michael Lingelbach ***@***.***> het volgende geschreven:
> 
> ﻿
> In that case it's extremely easy to override the default behavior with a fallback to util.path.dirname instead which I will document.
> 
> As it stands, people are putting util.path.dirname as a fallback for everything, which is not in compliance with the specification and will result in broken behavior (servers will fail to index the whole project) and getting confused when language servers do not work correctly. The root of the project is not/should not be the parent directory of the current file. We can special case a few (EFM, jsonls, for example).
> 

This sounds like a very legit issue, so I can align with the interest in wanting to address this.

Maybe having the UI picked as a non-default would suffice - it sounds like exactly what those users are looking for, and they're already overriding the defaults due to their unusual setup and lack of expertises in setting it up further. 

I personally don't have any overrides for finding the root directories: I've upstreamed all edge cases into this repo, hence why I'm not a fan of dropping these. They work in many common scenarios. Or sometimes it a matter of improving the LSP a bit. 

I think this is generally what we'd want to do: continuous refine these defaults so they work in as many scenarios as possible. Dropping these fallbacks kills room for further improvement.

**Comment by mjlbach** [comment]
> Maybe having the UI picked as a non-default would suffice - it sounds like exactly what those users are looking for, and they're already overriding the defaults due to their unusual setup and lack of expertises in setting it up further.

That defeats the purpose. You are the advanced user who is opting-in to having the language server attach to a buffer with an incorrect root (thus you should not expect the language server client to operate as designed). It should be trivial for you to change the fallback to `util.path.dirname`

>  They work in many common scenarios.  Or sometimes it a matter of improving the LSP a bit.

They don't, that's the issue, and I don't understand the second point.

> I think this is generally what we'd want to do: continuous refine these defaults so they work in as many scenarios as possible. Dropping these fallbacks kills room for further improvement.

That's exactly what I'm doing, the PR changes it such that the language server will only start when it will not error or fail to index your project files.

**Comment by mizlan** [comment]
Yeah, the assumption that the parent directory should "just work" for LSP is false. For editing files not in projects, wouldn't the normal assumption be to simply not enable LSP (maybe with a little reminder in the echo area)? Hence my suggestion above.

**Comment by WhyNotHugo** [comment]
> You are the advanced user who is opting-in to having the language server attach to a buffer with an incorrect root (thus you should not expect the language server client to operate as designed).

I'm the exact opposite of an advanced user; I've a very simple setup, sticking to upstream defaults as much as possible. I've no projects or codebases where the LSP misbehaves with these (though I'm sure they exist!).

If you have scenarios where they break, I'd advise researching them and figuring out how to improve the defaults this package provides. I don't think showing a UI to every single user every single time is a solution that makes sense: it promotes redoing the same work manually countless times instead of fixing how this package handles those scenarios. Users would be prompted to pick a root over and over and over and over.

Generally, showing a UI / prompt when starting up neovim is as invasive as a plugin can get, and something that, IMHO, should be avoided as much as possible. If I open a file, I want that to be opened, **not** to be distracted configuring an LSP. Like I said, I'm not an advanced user: if neovim distracts me with configuration questions, I'm most likely to lose forever whatever I was going to do.

---

This repo provides static configuration to rely on, and not things like prompts and wizards. I'd really prefer if it can stick to the purpose it has now, at least in its default behaviour.

**Comment by mjlbach** [comment]
> If you have scenarios where they break, I'd advise researching them and figuring out how to improve the defaults this package provides.

I have. The majority of the defaults are correct. You will almost never see the interactive prompt.

>  I don't think showing a UI to every single user every single time is a solution that makes sense: it promotes redoing the same work manually countless times instead of fixing how this package handles those scenarios. Users would be prompted to pick a root over and over and over and over.

You will almost never see the interactive prompt. It will only happen when we cannot automatically choose the root for you which will, as I said, rarely happen.

> Generally, showing a UI / prompt when starting up neovim is as invasive as a plugin can get, and something that, IMHO, should be avoided as much as possible. I

Then disable it, it will be a one line config option to do it.

**Comment by Hubro** [comment]
I feel like the most intuitive solution would be:

- If a proper project root directory can be resolved, enable LSP
- If not, don't enable LSP. If possible, display a small unintrusive message saying "LSP not enabled: A project root directory could not be resolved."

If the user wants to enable LSP anyway, they could run `:LspStart`. At this point, an interactive menu makes sense to me.

I don't think vim should ever "get in the way" by immediately displaying a prompt when the user opens a file.

**Comment by mjlbach** [comment]
@Hubro thanks for the feedback. I've come to a similar conclusion (mostly because of some annoyances I found regarding the timing of treesitter highlighting, which means the document is unhighlighted when the prompt is shown), however I would like to add as an option the automatic fallback.

As it stands, the user will be able to have two behaviors:
1. (default) the user will since a printed error that no root detected, and we can not automatically start the language server, at which point the user can call LspStart which will show the interactive root picker (which again, will check for a root, and only show the popup if no root is found)
2. The popup will be automatic if no root is found.
3. every single server defaults back to util.path.dirname, I will modify the issue template to make it clear this is unsupported and you must disable this when filing an issue
